File: code/javascript/data_structures_article-06.js
/**
 * Binary tree mental method (ideas)
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return flatten(...args);
}

var flatten = function(root) {
    if (root == null) return;
    flatten(root.left);
    flatten(root.right);
    var left = root.left;
    var right = root.right;
    root.left = null;
    root.right = left;
    var p = root;
    while (p.right != null) {
        p = p.right;
    }
    p.right = right;
};

File: code/python/data_structures_article-06.py
"""Binary tree mental method (ideas)"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().flatten(*args, **kwargs)


class _Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return
        self.flatten(root.left)
        self.flatten(root.right)
        left = root.left
        right = root.right
        root.left = None
        root.right = left
        p = root
        while p.right:
            p = p.right
        p.right = right
